--- a/iwinfo_cli.c
+++ b/iwinfo_cli.c
@@ -48,10 +48,21 @@ static char * format_channel(int ch)
 {
 	static char buf[8];
 
-	if (ch <= 0)
-		snprintf(buf, sizeof(buf), "unknown");
-	else
-		snprintf(buf, sizeof(buf), "%d", ch);
+
+	switch (ch) {
+		case 0:
+			snprintf(buf, sizeof(buf), "unknown");
+			break;
+		case 254:
+			snprintf(buf, sizeof(buf), "-2");
+			break;
+		case 255:
+			snprintf(buf, sizeof(buf), "-1");
+			break;
+       		default:
+			snprintf(buf, sizeof(buf), "%d", ch);
+			break;
+	}
 
 	return buf;
 }
@@ -409,7 +420,7 @@ static char * print_channel(const struct
 {
 	int ch;
 	if (iw->channel(ifname, &ch))
-		ch = -1;
+		ch = 0;
 
 	return format_channel(ch);
 }
@@ -644,7 +655,7 @@ static void print_freqlist(const struct
 	}
 
 	if (iw->channel(ifname, &ch))
-		ch = -1;
+		ch = 0;
 
 	for (i = 0; i < len; i += sizeof(struct iwinfo_freqlist_entry))
 	{
--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -518,12 +518,16 @@ static int nl80211_channel2freq(int chan
 	{
 		if (channel == 14)
 			return 2484;
+		else if (channel == 255)
+			return 2402;
+		else if (channel == 254)
+			return 2397;
 		else if (channel < 14)
 			return (channel * 5) + 2407;
 	}
 	else
 	{
-		if (channel >= 182 && channel <= 196)
+		if (channel >= 187 && channel <= 196)
 			return (channel * 5) + 4000;
 		else
 			return (channel * 5) + 5000;
